R/outpack_http_client.R

```diff
@@ -71,3 +87,28 @@ http_client_download_options <- function(dest) {
   c(httr::write_disk(dest),
     httr::accept("application/octet-stream"))
 }
+
+
+## Logging in with packit is quite slow and we'll want to cache this;
+## but we won't be holding a persistant handle to the root.  So for
+## now at least we'll keep a pool of generated bearer token headers,
+## stored against the hash of the auth details (so the url and the
+## token used to log in with).  We only store this on successful
+## login.
+##
+## This does mean there's no way to flush the cache and force a login,
+## but that should hopefully not be that big a problem.  We'll
+## probably want to refresh the tokens from the request anyway.
+auth_cache <- new.env(parent = emptyenv())
+http_client_login <- function(name, auth) {
+  key <- rlang::hash(auth)
+  if (is.null(auth_cache[[key]])) {
+    cli::cli_alert_info("Logging in to {name}")
+    res <- http_client_request(httr::POST, auth$url,
+                               body = auth$data, encode = "json")
+    cli::cli_alert_success("Logged in successfully")
+    auth_cache[[key]] <- httr::add_headers(
+      "Authorization" = paste("Bearer", res$token))
+  }
```

Comment:
Do we know how long the jwt tokens are valid for? If we get an error on request about an expired token we could clear the cache and retry the request with the token. Probably not a big issue, we might be unlikely to encounter it if the tokens are long living enough and easy to fix through manually clearing your cache or restarting the session.

Reply:
yeah, this is what I was thinking about with this comment:

> in orderlyweb (for orderly1) we refreshed the token on timeout; that can be handled fairly easily (see comment in http_client_handle_error)

This is how we handle it in orderlyweb: https://github.com/vimc/orderlyweb/blob/master/R/api_client.R#L111-L118

@EmmaLRussell ; do you know how Packit handles this? Or do you know how long the jwts are valid for?

Reply:
Jwt expiry time is configurable on deploy for packit: https://github.com/mrc-ide/packit/blob/main/api/app/src/main/resources/config.properties#L10

Reply:
amazing thanks - details now in mrc-5160 and will sort this out in another PR